### ISSUE TITLE
Received state for orders

### ISSUE THREAD

**Issue by lentschi** [foodcoops/foodsoft#779]
In order to facilitate "self-service" features such as https://github.com/foodcoops/foodsoft/issues/766 it would be nice to have a distinct "received" `state` in `Order`.

**Current status**
Currently it's hard to determine what (group) orders have been received - s. https://github.com/foodcoops/foodsoft/issues/766#issuecomment-710770954.

The order state flow is `open` -> `finished` -> `closed`.

**Planned adaptation**
The order state flow should become `open` -> `finished` -> `received` -> `closed`. All current queries checking for `Order.state = 'finished'` would then instead need to check for `Order.state IN ('finished', 'received')` (Though that may be subject to change in the future: For foodcoops in which "self service" is enabled - again see https://github.com/foodcoops/foodsoft/issues/766 - balancing may only be done for 'received' orders.)

**Comment by wvengen** [comment]
Sounds good!
Not all foodcoops use the receive screen, though, so it should be possible to skip that step (`finished` -> `closed`).

**Comment by lentschi** [comment]
Well this is exactly what I _intended_ to implement. :blush:  The `received` state is never set unless the users post the receive page.

You *can* still go from `finished` to `closed` in a single step. (Happens when settling the respective order for example - there is no check that the order must have been received first - see my original, self set goal: "All current queries checking for `Order.state = 'finished'` would then instead need to check for `Order.state IN ('finished', 'received')`")

Or am I missing some other foodsoft feature where this might be a problem? :thinking:

**Comment by wvengen** [comment]
Not that I can see. I'd be for implementing this.

It would be good to document this (e.g. in the model), including which state transitions can happen (to be sure other people realise what can be expected), and be sure to check all occurences of `finished`, also in plugins. And it would be good to mention it in the Changelog on the next release (for people with their own plugins).

**Comment by lentschi** [comment]
> It would be good to document this (e.g. in the model), including which state transitions can happen (to be sure other people realise what can be expected)

Yes, you're right! I've added it [here in my PR](https://github.com/foodcoops/foodsoft/pull/789/files#diff-26bbec4be8ce33f60fa824e2bd29e6902e57541818578634bea06bce180b620eR39).
Is that place fine?

> ... and be sure to check all occurences of finished, also in plugins.

I did. Scopes such as `Order.finished` already include the `received` state, so there's not so much that needed changing.
When going through the plugins, I found one deactivated plugin' controller where it may need to be changed:
https://github.com/foodcoops/foodsoft/blob/67ad20285933f77c7a961c96689591453d8283fe/plugins/current_orders/app/controllers/current_orders/group_orders_controller.rb#L7
However, I'm not sure how to display this page and the `current_orders` no longer seems to be working (even before my changes) - is it still maintained at all?  (For example [this count(distinct: true)](https://github.com/foodcoops/foodsoft/blob/67ad20285933f77c7a961c96689591453d8283fe/plugins/current_orders/app/views/current_orders/articles/_article.html.haml#L11) expression no longer seems to work in Rails 5 - it produces an SQL error.)

> And it would be good to mention it in the Changelog on the next release (for people with their own plugins).

Good point. Do you want me to add it directly to `CHANGELOG.md` with a `UPCOMING` heading? (I couldn't find anything about how to do this in the [Developing guidelines](https://github.com/foodcoops/foodsoft/wiki/Developing-Guidelines).)

If you do this manually once you release a new version - here's what I would put in there:

> Introduces the new optional order state `received` that is set after a user posts `OrdersController#receive`. The order state sequence thus becomes: `open -> finished (-> received) -> closed`. For most current foodsoft actions, `received` will be equivalent to `finished`. So plugin developers might need to alter their queries unless they really want to exclude orders that have been received! (The order scopes such as `Order.finished` *already* include the `received` state to facilitate this transition; only direct queries to `state` might need to be adapted.)

It's a bit verbose - so IMO you could shorten it, if you want to keep the Changelog's size low.
